### PR TITLE
Adding border_padding to montage2d

### PR DESCRIPTION
**Pull request by kmader**
## Description

A small tweak for montage2d to allow border padding: The `border_padding` argument adds a border around each tile in the montage so they can be more easily differentiated. Particularly helpful when looking at multidimensional weights where boundaries are not obvious.

### CODE REVIEW

File: skimage/util/montage.py [hunk 1]
```diff
@@ -78,11 +81,16 @@ def montage2d(arr_in, fill='mean', rescale_intensity=False, grid_shape=None):
     """
 
     assert arr_in.ndim == 3
-
+    
+    # -- add border padding (np.pad does all dimensions 
+    # so we remove the padding from the first
+    if border_padding>0:
```
Comment (sciunto) [inline skimage/util/montage.py]: spaces around >

File: skimage/util/montage.py [hunk 2]
```diff
@@ -6,7 +6,7 @@
 EPSILON = 1e-6
 
 
-def montage2d(arr_in, fill='mean', rescale_intensity=False, grid_shape=None):
+def montage2d(arr_in, fill='mean', rescale_intensity=False, grid_shape=None, border_padding = 0):
```
Comment (sciunto) [inline skimage/util/montage.py]: PEP8: no spaces around = sign.

File: skimage/util/montage.py [hunk 3]
```diff
@@ -78,11 +81,16 @@ def montage2d(arr_in, fill='mean', rescale_intensity=False, grid_shape=None):
     """
 
     assert arr_in.ndim == 3
-
+    
+    # -- add border padding (np.pad does all dimensions
```
Comment (sciunto) [inline skimage/util/montage.py]: unbalanced parenthesis.

**Comment by sciunto** [comment]
Thanks @kmader I left few comments to fix on the style.

**Comment by sciunto** [comment]
Looks good to me. Let's wait for a second review :)

File: skimage/util/montage.py [hunk 4]
```diff
@@ -6,7 +6,7 @@
 EPSILON = 1e-6
 
 
-def montage2d(arr_in, fill='mean', rescale_intensity=False, grid_shape=None):
```
Comment (soupault) [inline skimage/util/montage.py]: I'd rename this parameter as 'padding_width' or smth like that. `border_padding` sounds like something binary (yes/no).

File: skimage/util/montage.py [hunk 5]
```diff
@@ -78,11 +81,16 @@ def montage2d(arr_in, fill='mean', rescale_intensity=False, grid_shape=None):
     """
 
     assert arr_in.ndim == 3
-
+    
+    # -- add border padding, np.pad does all dimensions 
+    # so we remove the padding from the first
+    if border_padding > 0:
+        arr_in = np.pad(arr_in, border_padding, mode='constant')[border_padding:-border_padding]
```
Comment (soupault) [inline skimage/util/montage.py]: Why not just wisely use the 2nd argument so to not allocate extra memory?

File: skimage/util/montage.py [hunk 6]
```diff
@@ -28,21 +28,24 @@ def montage2d(arr_in, fill='mean', rescale_intensity=False, grid_shape=None):
...
         is not an integer. If 'mean' is chosen, then fill = arr_in.mean().
-    rescale_intensity: bool, optional
+    rescale_intensity : bool, optional
         Whether to rescale the intensity of each image to [0, 1].
-    grid_shape: tuple, optional
+    grid_shape : tuple, optional
         The desired grid shape for the montage (tiles_y, tiles_x).
         The default aspect ratio is square.
+    border_padding : int, optional
+        The size of the spacing between the tiles to make the
```
Comment (soupault) [inline skimage/util/montage.py]: This is not exactly true, as the spacing between the tiles will be twice as much.

**Comment by soupault** [comment]
@kmader thanks! This will also require some testing in `test_montage` (couple of simple tests will do the deal).

File: skimage/util/montage.py (hunk 5 above)
Comment (emmanuelle) [inline skimage/util/montage.py]: As mentioned by @soupault , ``pad_width`` in ``np.pad`` can be an iterable, so you can specify 0-padding along the 1st dimension inside ``np.pad``.

File: skimage/util/montage.py [hunk 7]
```diff
@@ -108,3 +118,84 @@ def montage2d(arr_in, fill='mean', rescale_intensity=False, grid_shape=None):
     arr_out = arr_out.reshape(alpha_y * height, alpha_x * width)
 
     return arr_out
+
+def montage_rgb(arr_in, fill='mean', grid_shape=None, border_padding=0):
```
Comment (emmanuelle) [inline skimage/util/montage.py]: Do you really need this function? I have the feeling that there is a lot of duplicated code from ``montage2d``. Can't you just build the montage for each channel and then combine them together?

**Comment by emmanuelle** [comment]
@kmader thanks for your contribution. Do you have time to address the comments that were made?

**Comment by ahojnnes** [comment]
@kmader Thanks. This looks good to me, but we would need unit tests for the new functionality.

**Comment by ahojnnes** [comment]
@kmader Tests look great. One last nitpicky comment: please, add two empty lines between functions to comply with PEP008. Thanks!

**Comment by sciunto** [comment]
You can add your name in the contributors list, if you wish.

File: skimage/util/montage.py (hunk 7 above)
Comment (soupault) [inline skimage/util/montage.py]: I agree on this point. `montage2d` and `montage_rgb` perform, basically, the same operation. Could be better to integrate the latter into the former, and add simple check on `ndim` of the input.
Although, if we'd are planning to have `montage_channels` function (which will create a matrix of different channels, not images), we shouldn't merge this functions.

File: skimage/util/montage.py [hunk 8]
```diff
@@ -108,3 +122,67 @@ def montage2d(arr_in, fill='mean', rescale_intensity=False, grid_shape=None):
...
+    
+    n_images, height, width, n_channels = arr_in.shape
+    
+    out_slices = []
+    for i_chan in range(n_channels):
+        out_slices += [montage2d(arr_in[:,:,:,i_chan], fill=fill, 
+                                 grid_shape=grid_shape, padding_width=padding_width)]
+    
+
+    return np.stack(out_slices,2)
```
Comment (emmanuelle) [inline skimage/util/montage.py]: please add space after comma for PEP8

**Comment by emmanuelle** [comment]
Hi @kmader thanks for your changes, that was quick! It seems that there are conflicts with the master branch. Can you please try rebasing to see if you can resolve the conflicts? (see http://scikit-image.org/docs/stable/contribute.html#divergence-between-upstream-master-and-your-feature-branch).

**Comment by emmanuelle** [comment]
ping @kmader : I'm almost sure your PR is ready for merging, but please rebase so that we can be sure (if you need any help for rebasing, please tell us so).

**Comment by kmader** [comment]
@emmanuelle sorry for the delay and thanks for the rebasing instructions, everything should be fine now

**Comment by sciunto** [comment]
Continued in #2626 Thanks @kmader

**Pull request closed without merging**